content/blog/2019-11-13-trace-gen-llvm/index.md

```diff
@@ -0,0 +1,159 @@
+
+Let's generate traces from LLVM!
+---------------------------
+
+The code described in this blog post is hosted [here](https://github.com/pbb59/ir-trace).
+
+## Overview
+
+In this project we use LLVM to generate traces and execute them. We use LLVM passes to log all the traces in a program and use another pass to create an executable that only executes a selected trace. This could be useful to implement efficient trace selection, or to implement trace-based LLVM passes in general.
+
+In a real system, an interpreter would internally implement both of these passes (profiling and trace generation). However, we did not use an interpreter due to technological limitations in LLVM.
+
+## Background
+
+Trace-based compilers have found many use cases in computer systems. In tracing compilers, A piece of a program (generally a loop body) is compiled to only execute a single path through that piece as shown below. 
+
+<img src="tracing-jit.png" alt="Tracing" width="20%">
+
+Guards must be inserted to check whether this assumption of path taken by the program is correct. If it's not correct, then another path must be executed at high performance overhead. These systems speculate on a certain control flow of the program in a somewhat high-risk high-reward environment: the benefits of repeating the same trace is large (fewer instructions and more optimizations between basic blocks), but mispredicting the trace is worse than not running the trace at all.
+
+Tracing systems found early success in VLIW architectures. These architectures try to schedule many instructions at once, but run into trouble scheduling across basic blocks, which may or may not be run. A single basic block can remedy this by only considering instructions along a single path through the program.
+
+More recently, tracing has found success in Just-in-Time compilers (JITs) for dynamic languages. The most popular Python JIT is a tracing JIT that speculates on both types and control flow.
+
+## Trace Detection
+
+To determine all the traces in a given program, we first use LLVM functions to detect branch instructions. We isolate branch instructions that diverge control flow and then insert a function call to a runtime library to log the condition upon divergence. At runtime, when the conditions to the diverging branch instructions are resolved, the sequence of branch selection at execution is generated as a CSV file of boolean values. This sequence of branch selection forms a trace through the entire program taken at the execution of trace detection runtime.
+
+This LLVM pass was created broadly following the "Linking With a Runtime Library" section from [LLVM for Grad Students](https://www.cs.cornell.edu/~asampson/blog/llvm.html) blog post. However, a `Value` was created from `FunctionCallee` object before passing into `CreateCall` function from [`IRBuilder`](https://llvm.org/doxygen/classllvm_1_1IRBuilder.html) to avoid a casting error. This pass is registered as `-skull` in `Skull.cpp`.
```

Comment:
No need to delete this or anything, but this level of implementation detail is probably a tad too granular for the goal here—we don't need to know exactly which lines of code you wrote to understand that you inserted calls to a runtime library.